ignore measurements outside lidar range

981213 opened the pull request:
According to the comment in sensor_msgs/LaserScan, values < range_min or > range_max should be discarded.
Without this, scan matching will fail for lidars publishing 0 for invalid measurements.

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Primary OS tested on | Ubuntu 22.04 / ROS2 Humble |
| Robotic platform tested on | My DIY robot with LeiShen Lidar |

---

## Description of contribution in a few bullet points

* Don't add a point to UnfilteredPoingReadings when the range reading is outside lidar range.

SteveMacenski commented on lib/karto_sdk/include/karto_sdk/Karto.h:
```diff
@@ -5658,27 +5658,21 @@ class LocalizedRangeScan : public LaserRangeScan
-
-          m_UnfilteredPointReadings.push_back(point);
+        if (!math::InRange(rangeReading, pLaserRangeFinder->GetMinimumRange(),
+                           pLaserRangeFinder->GetMaximumRange()))
```
https://github.com/SteveMacenski/slam_toolbox/blob/2f869c05a8047dc3c54b2bc12e52b6fb39d69d40/lib/karto_sdk/include/karto_sdk/Karto.h#L3948

The `rangeThreshold` is always less than `GetMaximumRange` so this is already true. You can revert these changes

SteveMacenski commented on lib/karto_sdk/include/karto_sdk/Karto.h:
```diff
@@ -5658,27 +5658,21 @@ class LocalizedRangeScan : public LaserRangeScan
 
-        m_PointReadings.push_back(point);
+        if (rangeReading < rangeThreshold) {
+            m_PointReadings.push_back(point);
```
I believe that the occupancy grid generator needs the number of point readings to be the same as the number coming in since it assumes a regularized angle of each laser beam. If you drop points, then I think that will mess up the scan matching and/or the occupancy grid generation.

SteveMacenski replied:
Can you test if you remove some large number of scan points that the output map still looks correct  (occ grid generator) & that scan matching still works properly (localization quality over large maps still works well)?

981213 replied:
It does work, but with my setup and environment, this patch drops fixed number of points (beams where the lidar is covered by my robot body), and it doesn't break the assumption that the angle of each beam is fixed.
I'll change it to push NaN points into the array instead of dropping it.

SteveMacenski commented on lib/karto_sdk/include/karto_sdk/Karto.h:
```diff
@@ -5658,7 +5658,10 @@ class LocalizedRangeScan : public LaserRangeScan
       for (kt_int32u i = 0; i < pLaserRangeFinder->GetNumberOfRangeReadings(); i++, beamNum++) {
         kt_double rangeReading = GetRangeReadings()[i];
-        if (!math::InRange(rangeReading, pLaserRangeFinder->GetMinimumRange(), rangeThreshold)) {
+        if (rangeReading < pLaserRangeFinder->GetMinimumRange()) {
```
Why not using that `InRange` line that was there before `math::InRange(rangeReading, pLaserRangeFinder->GetMinimumRange(), rangeThreshold)`? This seems to remove the far-check

981213 committed:
Push NAN into point array for measurement outside range limit

According to the comment in sensor_msgs/LaserScan, values
< range_min or > range_max should be discarded.
Without this, scan matching will fail for lidars publishing
0 for invalid measurements.

981213 commented:
I moved the filtration outside karto_sdk in the latest iteration. The scan matching is fixed in both the old and this version, but loop closure still doesn't seem right with my half-blocked lidar.

I placed a second lidar on top of the half-blocked one to do some comparison. Here's my robot setup:
![Robot Setup](https://user-images.githubusercontent.com/3888062/229714708-79bde866-b07f-44cb-988b-12fb6fdffa6f.jpg)
With async slam node, the top lidar and all the default configuration, it works perfectly:
![Result with the top LSN10 lidar](https://user-images.githubusercontent.com/3888062/229715091-3badf7c0-1612-4442-a23f-8ca4f9f29ca4.png)
But here's the result using the bottom one:
![Screenshot_20230404_145322](https://user-images.githubusercontent.com/3888062/229715449-3c1e022c-0f5c-487b-839f-0c06cf890054.png)
I wrote the lidar driver myself: https://github.com/981213/ls01_ros2 and the blocked measurements are set to 0 with memset in my code.

It would be great if someone can give me some suggestions on debugging.
Here's the ros2 bag for the test above.
[rosbag2_2023_04_04-14_16_58.zip](https://github.com/SteveMacenski/slam_toolbox/files/11145673/rosbag2_2023_04_04-14_16_58.zip)

/scan is the half-blocked lidar and /scan2 is the one on top.
The missing lidar transform:
```
ros2 run tf2_ros static_transform_publisher 0.03 0.0 0.1415 0 0 0 base_footprint laser_link
```

981213 commented:
Fixed with the latest commit, but I'm not sure if it's the correct fix because I haven't actually understood how ScanMatcher works yet.

edit: dropped this. I should simply lower the threshold in config file instead.
